File: libgomp/testsuite/libgomp.oacc-c-c++-common/private-vars-par-gang-2.c
/* { dg-additional-options "-w" } */

#include <assert.h>
#include <openacc.h>

/* Test of gang-private variables declared on the parallel directive.  */

#define ACTUAL_GANGS 32

int
main (int argc, char* argv[])
{
  int x = 5, i, arr[ACTUAL_GANGS];

  for (i = 0; i < ACTUAL_GANGS; i++)
    arr[i] = 3;

  #pragma acc parallel private(x) copy(arr) num_gangs(ACTUAL_GANGS) \
		       num_workers(8) vector_length(32)
  {
    #pragma acc loop gang(static:1)
    for (i = 0; i < ACTUAL_GANGS; i++)
      x = i * 2;

    #pragma acc loop gang(static:1)
    for (i = 0; i < ACTUAL_GANGS; i++)
      {
	if (acc_on_device (acc_device_host))
	  x = i * 2;
	arr[i] += x;
      }
  }

  for (i = 0; i < ACTUAL_GANGS; i++)
    assert (arr[i] == 3 + i * 2);

  return 0;
}

File: libgomp/testsuite/libgomp.oacc-fortran/private-vars-par-gang-2.f90
! { dg-additional-options "-w" }

! Test of gang-private variables declared on the parallel directive.

program main
  use openacc
  integer :: x = 5
  integer, parameter :: n = 32
  integer :: arr(n)

  do i = 1, n
    arr(i) = 3
  end do

  !$acc parallel private(x) copy(arr) num_gangs(n) num_workers(8) vector_length(32)
    !$acc loop gang(static:1)
    do i = 1, n
      x = i * 2;
    end do

   !$acc loop gang(static:1)
    do i = 1, n
      if (acc_on_device (acc_device_host) .eqv. .TRUE.) x = i * 2
      arr(i) = arr(i) + x
    end do
  !$acc end parallel

  do i = 1, n
    if (arr(i) .ne. (3 + i * 2)) call abort
  end do

end program main
